man/pipes.3
.TH PIPES 3 2014-09-03 "" "PIPES"
.SH NAME
pipes \- build pipe chains

.SH SYNOPSIS
.br
.nf
.B #include <pipes/pipes.h>

.SS "Data Structures"
.nf
struct \fBpipes\fP;
struct \fBpipes_chain\fP;

.SS "Functions"
.nf
int \fBpipes_open\fP(char const *const \fIargv\fP[], char const *const \fIenvp\fP[],
               struct \fBpipes\fP* \fIpipes\fP);
int \fBpipes_close\fP(struct \fBpipes\fP* \fIpipes\fP);
.sp
int \fBpipes_open_chain\fP(struct \fBpipes_chain\fP \fIchain\fP[]);
int \fBpipes_close_chain\fP(struct \fBpipes_chain\fP \fIchain\fP[]);
int \fBpipes_kill_chain\fP(struct \fBpipes_chain\fP \fIchain\fP[], int \fIsig\fP);
.sp
int \fBpipes_take_in\fP(struct \fBpipes_chain\fP \fIchain\fP[]);
int \fBpipes_take_out\fP(struct \fBpipes_chain\fP \fIchain\fP[]);
int \fBpipes_take_err\fP(struct \fBpipes_chain\fP \fIchain\fP[]);

.SS "Macros"
.nf
#define \fBPIPES_GET_LAST\fP(\fICHAIN\fP)
#define \fBIPES_GET_IN\fP(\fICHAIN\fP)
#define \fBIPES_GET_OUT\fP(\fICHAIN\fP)
#define \fBIPES_GET_ERR\fP(\fICHAIN\fP)

.SH DESCRIPTION
Functions to build pipe chains between programs.

.SS sctruct pipes

.PP
.in +4n
.nf
struct pipes {
	pid_t pid;     /* process ID of opened child process */
	int   infd;    /* pipe to stdin of child process     */
	int   outfd;   /* pipe to stdout of child process    */
	int   errfd;   /* pipe to stderr of child process    */
};
.fi
.in

The \fIinfd\fP, \fIoutfd\fP, and \fIerrfd\fP fields of \fIpipes\fP may either be set to valid
file descriptors or to one of the following special values. In case they are file descriptors
they will be used as the appropriate stream in the child process and will be closed in the
parent process. \fBpipes_open\fP() and \fBpipes_open_chain\fP() ensure that any passed file
descriptor will be always closed, no matter if the call was successful or resulted in an error.

.TP
.B PIPES_LEAVE
Leave stream unchaned (i.e. don't open a pipe to the stream of the child process).

.TP
.B PIPES_PIPE
Open a pipe to this stream of the child process. For \fIinfd\fP the write end of the pipe will
be returned in the same field. For \fIoutfd\fP and \fIerrfd\fP the read end of the pipe will
be retuned in the same field.

.TP
.B PIPES_NULL
Pipe this stream to/from \fI/dev/null\fP.

.TP
.B PIPES_ERR_TO_OUT
Redirect child process stderr to stdout. This is only valid for the \fIerrfd\fP field.

.TP
.B PIPES_TEMP
Open a temporary file as this stream. If supported by the kernel and filesystem this
will be done using the \fBO_TMPFILE\fP flag. A copy of the file descriptor will be
returned in the same field.

.PP
These helper macros can be used to initialize the \fIpipes\fP structure:

.TP
.B PIPES_PASS
Set \fIinfd\fP and \fIoutfd\fP to \fBPIPES_PIPE\fP and set \fIerrfd\fP to \fBPIPES_LEAVE\fP.
This is useful for intermediaries in a pipe chain.

.TP
.BI PIPES_IN( IN )
Set \fIinfd\fP to \fIIN\fP, \fIoutfd\fP to \fBPIPES_PIPE\fP and \fIerrfd\fP to
\fBPIPES_LEAVE\fP.

.TP
.BI PIPES_OUT( OUT )
Set \fIinfd\fP to \fBPIPES_PIPE\fP, \fIoutfd\fP to \fIOUT\fP and \fIerrfd\fP to
\fBPIPES_LEAVE\fP.

.TP
.BI PIPES_ERR( ERR )
Set \fIinfd\fP to \fBPIPES_PIPE\fP, \fIoutfd\fP to \fBPIPES_LEAVE\fP and \fIerrfd\fP to
\fIERR\fP.

.TP
.B PIPES_FIRST
Set \fIoutfd\fP to \fBBR PIPES_PIPE\fP and the rest to \fBPIPES_LEAVE\fP. This is useful for
the first command in a chain.

.TP
.B PIPES_LAST
Set \fIinfd\fP to \fBPIPES_PIPE\fP and the rest to \fBPIPES_LEAVE\fP. This is useful for the
last command in a chain.

.SS struct pipes_chain

.PP
.in +4n
.nf
struct pipes_chain {
	struct pipes       pipes;   /* see above                         */
	char const* const* argv;    /* NULL terminated argument array    */
	char const* const* envp;    /* NULL terminated environment array */
};
.fi
.in

\fBpipes_chain_open\fP() accepts an array of \fBpipe_chain\fP structures. It passed the fields
of each structure to an \fBpipes_open\fP() call.

.SS int pipes_open(char const *const \fIargv\fP[], char const *const \fIenvp\fP[], struct pipes* \fIpipes\fP);
Spawn a child process and open pipes to it's io streams.

\fIargv\fP is a NULL terminated array of arguments. The first one is the program to execute
and does not need to be an absolute path.

\fIenvp\fP is a NULL terminated array of environment variables. After forking, the child
processes global \fBenviron\fP variable will be set to this value. If \fIenvp\fP is NULL it will
be ignored. See also: \fBenviron\fP(3)

On success returns 0, on error returns -1 and sets \fBerrno\fP. If \fIinfd\fP, \fIoutfd\fP or
\fIerrfd\fP has an illegal value \fBerrno\fP is set to \fBEINVAL\fP. For other possible error
codes see \fBopen\fP(2), \fBpipe\fP(2), \fBdup2\fP(2), and \fBfork\fP(2).

.SS int pipes_close(struct pipes* \fIpipes\fP)
Close all pipes previously opened with \fBpipes_open\fP(). It is save to call this even if the
\fBpipes_open\fP() call failed.

Returns 0 on success, -1 if \fBclose(2)\fP on any of the file descriptors failed.

.SS int pipes_open_chain(struct pipes_chain \fIchain\fP[])
Spawn a number of child prcesses and open pipes between them. Intermediate pipes are
not accessible by the calling process.

The last element of \fIchain\fP is marked by setting \fIargv\fP to NULL. \fIenvp\fP can be NULL
and \fIpipes\fP must be initialized in the same way as for \fBpipes_open\fP().

On success returns 0, on error returns -1 and sets \fBerrno\fP. In addition to the errors
of \fBopen_pipes\fP()

.SS pipes_close_chain(struct pipes_chain \fIchain\fP[])
Close all pipes in \fIchain\fP. It is save to call this even if the \fBpipes_open\fP() call
failed.

.SS pipes_kill_chain(struct pipes_chain \fIchain\fP[], int \fIsig\fP)
Send signal \fIsig\fP to all processes in \fIchain\fP.

.SS pipes_take_in(struct pipes_chain \fIchain\fP[])
Return the pipe to the input stream pipe of the first process in the \fIchain\fP. The \fIinfd\fP
field in the chain will be set to -1 so a successive \fBpipes_close_chain\fP() call won't close
it.

.SS pipes_take_out(struct pipes_chain \fIchain\fP[])
Return the pipe to the output stream pipe of the last process in the \fIchain\fP. The \fIoutfd\fP
field in the chain will be set to -1 so a successive \fBpipes_close_chain\fP() call won't close it.

.SS pipes_take_err(struct pipes_chain \fIchain\fP[])
Return the pipe to the error stream pipe of the last process in the \fIchain\fP. The \fIerrfd\fP
field in the chain will be set to -1 so a successive \fBpipes_close_chain\fP() call won't close it.

.SS PIPES_GET_LAST(\fICHAIN\fP)
Macro to get the last pipe in \fICHAIN\fP. \fICHAIN\fP must be an array, not a pointer.

.SS PIPES_GET_IN(\fICHAIN\fP)
Macro to get the input stream pipe of the first process in \fICHAIN\fP.

.SS PIPES_GET_OUT(\fICHAIN\fP)
Macro to get the output stream pipe of the last process in \fICHAIN\fP. \fICHAIN\fP must be an
array, not a pointer.

.SS PIPES_GET_ERR(\fICHAIN\fP)
Macro to get the error stream pipe of the last process in \fICHAIN\fP. \fICHAIN\fP must be an
array, not a pointer.

.SH ERRORS
TODO...

.TP
.B EINVAL
TODO...

.SH SOURCE
Get the source at https://github.com/panzi/pipes

.SH AUTHOR
Mathias Panzenböck (grosser.meister.morti@gmx.net)

.SH SEE ALSO
.BR fpipes (3),
.BR environ (3),
.BR execvp (3),
.BR fork (3),
.BR popen (3)


pipes.c
#define _POSIX_SOURCE
#define _GNU_SOURCE

#include "pipes.h"

#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>

#ifdef __APPLE__
#	include <crt_externs.h>
#	define environ (*_NSGetEnviron())
#else
	extern char **environ;
#endif

void pipes_redirect_fd(int oldfd, int newfd, const char *msg);

#ifndef P_tmpdir
#	define P_tmpdir "/tmp"
#endif

static int pipes_temp_fd_fallback() {
	char name[L_tmpnam];

	if (tmpnam(name) == NULL) {
		return -1;
	}

	int fd = open(name, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);

	if (fd < 0) {
		return -1;
	}

	if (unlink(name) != 0) {
		close(fd);
		return -1;
	}

	return fd;
}

#if defined(O_TMPFILE) || defined(__linux__)
#	ifndef O_TMPFILE
#		define O_TMPFILE (020000000 | O_DIRECTORY)
#	endif
static int pipes_temp_fd() {
	int fd = open(P_tmpdir, O_TMPFILE | O_RDWR, S_IRUSR | S_IWUSR);
	if (fd < 0 && (
		errno == EOPNOTSUPP || // no filesystem support for O_TMPFILE
		errno == EISDIR     || // no kernel support for O_TMPFILE
		errno == ENOENT)) {    // no kernel support for O_TMPFILE and path does not exist
		return pipes_temp_fd_fallback();
	}
	return fd;
}
#else
#	define pipes_temp_fd() pipes_temp_fd_fallback()
#endif

int pipes_open(char const *const argv[], char const *const envp[], struct pipes* pipes) {
	int infd  = -1;
	int outfd = -1;
	int errfd = -1;

	const int inaction  = pipes->infd;
	const int outaction = pipes->outfd;
	const int erraction = pipes->errfd;

	// stdin
	if (inaction == PIPES_PIPE) {
		int pair[] = {-1, -1};
		if (pipe(pair) == -1) {
			goto error;
		}

		infd = pair[0];
		pipes->infd = pair[1];
	}
	else if (inaction == PIPES_NULL) {
		infd = open("/dev/null", O_RDONLY);

		if (infd < 0) {
			goto error;
		}
	}
	else if (inaction > -1) {
		infd = inaction;
		pipes->infd = -1;
	}
	else if (inaction == PIPES_TEMP) {
		infd = pipes_temp_fd();
		pipes->infd = infd;

		if (infd < 0) {
			goto error;
		}
	}
	else if (inaction != PIPES_LEAVE) {
		errno = EINVAL;
		goto error;
	}

	// stdout
	if (outaction == PIPES_PIPE) {
		int pair[] = {-1, -1};
		if (pipe(pair) == -1) {
			goto error;
		}

		pipes->outfd = pair[0];
		outfd = pair[1];
	}
	else if (outaction == PIPES_NULL) {
		outfd = open("/dev/null", O_WRONLY);

		if (outfd < 0) {
			goto error;
		}
	}
	else if (outaction > -1) {
		outfd = outaction;
		pipes->outfd = -1;
	}
	else if (outaction == PIPES_TEMP) {
		outfd = pipes_temp_fd();
		pipes->outfd = outfd;

		if (outfd < 0) {
			goto error;
		}
	}
	else if (outaction != PIPES_LEAVE) {
		errno = EINVAL;
		goto error;
	}

	// stderr
	if (erraction == PIPES_PIPE) {
		int pair[] = {-1, -1};
		if (pipe(pair) == -1) {
			goto error;
		}

		pipes->errfd = pair[0];
		errfd = pair[1];
	}
	else if (erraction == PIPES_NULL) {
		errfd = open("/dev/null", O_WRONLY);

		if (errfd < 0) {
			goto error;
		}
	}
	else if (erraction == PIPES_ERR_TO_OUT) {
		// see below
	}
	else if (erraction > -1) {
		errfd = erraction;
		pipes->errfd = -1;
	}
	else if (erraction == PIPES_TEMP) {
		errfd = pipes_temp_fd();
		pipes->errfd = errfd;

		if (errfd < 0) {
			goto error;
		}
	}
	else if (erraction != PIPES_LEAVE) {
		errno = EINVAL;
		goto error;
	}

	pid_t pid = fork();

	if (pid == -1) {
		goto error;
	}

	if (pid == 0) {
		// child
		pipes_redirect_fd(infd,  STDIN_FILENO,  "redirecting stdin");
		pipes_redirect_fd(outfd, STDOUT_FILENO, "redirecting stdout");
		if (erraction == PIPES_ERR_TO_OUT) {
			if (dup2(STDOUT_FILENO, STDERR_FILENO) == -1) {
				perror("redirecting stderr");
				exit(EXIT_FAILURE);
			}
		}
		else {
			pipes_redirect_fd(errfd, STDERR_FILENO, "redirecting stderr");
		}

		if (envp) {
			environ = (char**)envp;
		}

		if (execvp(argv[0], (char * const*)argv) == -1) {
			perror(argv[0]);
		}
		exit(EXIT_FAILURE);
	}
	else {
		// parent
		pipes->pid = pid;

		if (inaction  != PIPES_TEMP && infd  > -1) close(infd);
		if (outaction != PIPES_TEMP && outfd > -1) close(outfd);
		if (erraction != PIPES_TEMP && errfd > -1) close(errfd);
	}

	return 0;

error:
	pipes->pid = -1;

	int errnum = errno;

	if (infd  > -1) close(infd);
	if (outfd > -1) close(outfd);
	if (errfd > -1) close(errfd);

	pipes_close(pipes);

	if (errnum != 0) {
		errno = errnum;
	}

	return -1;
}

int pipes_close(struct pipes* pipes) {
	int status = 0;

	if (pipes->infd  > -1) {
		if (close(pipes->infd) != 0) {
			status = -1;
		}
		pipes->infd = -1;
		status = -1;
	}

	if (pipes->outfd > -1) {
		if (close(pipes->outfd) != 0) {
			status = -1;
		}
		pipes->outfd = -1;
	}

	if (pipes->errfd > -1) {
		if (close(pipes->errfd) != 0) {
			status = -1;
		}
		pipes->errfd = -1;
	}

	return status;
}

int pipes_open_chain(struct pipes_chain chain[]) {
	struct pipes_chain *ptr  = chain;
	struct pipes_chain *prev = chain;

	for (; ptr->argv; ++ ptr) {
		ptr->pipes.pid = -1;
	}

	if (chain == NULL || chain[0].argv == NULL) {
		errno = EINVAL;
		goto error;
	}

	ptr = chain;

	for (++ ptr; ptr->argv; ++ ptr) {
		if (ptr->pipes.infd == PIPES_PIPE && prev->pipes.outfd != PIPES_PIPE) {
			errno = EINVAL;
			goto error;
		}
		prev = ptr;
	}

	ptr  = chain;
	prev = chain;

	if (pipes_open(ptr->argv, ptr->envp, &ptr->pipes) == -1) {
		goto error;
	}

	for (++ ptr; ptr->argv; ++ ptr) {
		if (ptr->pipes.infd == PIPES_PIPE) {
			ptr->pipes.infd   = prev->pipes.outfd;
			prev->pipes.outfd = -1;
		}

		if (pipes_open(ptr->argv, ptr->envp, &ptr->pipes) == -1) {
			goto error;
		}

		prev = ptr;
	}

	return 0;

error:

	(void)0;

	int errnum = errno;

	pipes_close_chain(chain);
	pipes_kill_chain(chain, SIGTERM);

	if (errnum != 0) {
		errno = errnum;
	}

	return -1;
}

int pipes_close_chain(struct pipes_chain chain[]) {
	int status = 0;

	for (struct pipes_chain *ptr = chain; ptr->argv; ++ ptr) {
		if (pipes_close(&ptr->pipes) != 0) {
			status = -1;
		}
	}

	return status;
}

int pipes_kill_chain(struct pipes_chain chain[], int sig) {
	int status = 0;

	for (struct pipes_chain *ptr = chain; ptr->argv; ++ ptr) {
		if (ptr->pipes.pid > -1) {
			if (kill(ptr->pipes.pid, sig) != 0) {
				status = -1;
			}
		}
	}

	return status;
}

int pipes_take_in(struct pipes_chain chain[]) {
	if (chain[0].argv) {
		int fd = chain[0].pipes.infd;
		chain[0].pipes.infd = -1;
		return fd;
	}
	else {
		errno = EINVAL;
		return -1;
	}
}

int pipes_take_out(struct pipes_chain chain[]) {
	struct pipes_chain *prev = chain;
	for (struct pipes_chain *ptr = chain; ptr->argv; ++ ptr) {
		prev = ptr;
	}

	if (prev->argv) {
		int fd = prev->pipes.outfd;
		prev->pipes.outfd = -1;
		return fd;
	}
	else {
		errno = EINVAL;
		return -1;
	}
}

int pipes_take_err(struct pipes_chain chain[]) {
	struct pipes_chain *prev = chain;
	for (struct pipes_chain *ptr = chain; ptr->argv; ++ ptr) {
		prev = ptr;
	}

	if (prev->argv) {
		int fd = prev->pipes.errfd;
		prev->pipes.errfd = -1;
		return fd;
	}
	else {
		errno = EINVAL;
		return -1;
	}
}
